File: src/main/java/org/beatonma/lib/ui/activity/BaseFragment.java
package org.beatonma.lib.ui.activity;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.material.snackbar.Snackbar;

import java.lang.ref.WeakReference;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.fragment.app.Fragment;

/**
 * Created by Michael on 05/08/2016.
 */
@SuppressWarnings("unused")
public abstract class BaseFragment extends Fragment {
    protected final static String TAG = "BaseFragment";
    protected WeakReference<Context> mWeakContext;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mWeakContext = new WeakReference<>(getContext());
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater,
                             @Nullable final ViewGroup container,
                             @Nullable final Bundle savedInstanceState) {
        final ViewDataBinding binding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false);

        init(binding);
        postInit();

        return binding.getRoot();
    }

    protected abstract void init(ViewDataBinding binding);
    protected abstract int getLayoutId();

    @CallSuper
    public void postInit() {

    }

    public BaseActivity getParentActivity() {
        final Activity activity = getActivity();
        if (activity instanceof BaseActivity) {
            return (BaseActivity) activity;
        }
        throw new ClassCastException("Parent activity does not extend BaseActivity");
    }

    public void showSnackbar(final String message) {
        final Activity activity = getActivity();
        if (activity instanceof BaseActivity) {
            ((BaseActivity) activity).showSnackbar(message);
        }
        else {
            Log.e(TAG, "Cannot show snackbar: activity does not extend BaseActivity");
        }
    }

    public Snackbar getSnackbar(final String message) {
        final Activity activity = getActivity();
        if (activity instanceof BaseActivity) {
            return ((BaseActivity) activity).getSnackbar(message);
        }
        else {
            Log.e(TAG, "Cannot show snackbar: activity does not extend BaseActivity");
            return null;
        }
    }

    public View getSharedViewStub() {
        return getParentActivity().getSharedViewStub();
    }
}

File: src/main/java/org/beatonma/lib/ui/activity/BaseFragment.kt
package org.beatonma.lib.ui.activity

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.core.view.doOnPreDraw
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import com.google.android.material.snackbar.Snackbar

/**
 * Created by Michael on 05/08/2016.
 */
private const val TAG = "BaseFragment"
abstract class BaseFragment : Fragment() {
    protected abstract val layoutID: Int

    val parentActivity: BaseActivity
        get() = activity as? BaseActivity
                ?: throw ClassCastException("Parent activity does not extend BaseActivity")
    val sharedViewStub: View?
        get() = parentActivity.sharedViewStub

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        val binding = DataBindingUtil.inflate<ViewDataBinding>(inflater, layoutID, container, false)

        init(binding)

        return binding.root
    }


    @CallSuper
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        postponeEnterTransition()

        (view.parent as? ViewGroup)?.doOnPreDraw {
            startPostponedEnterTransition()
        }
    }

    protected abstract fun init(binding: ViewDataBinding)

    fun showSnackbar(message: String) {
        (activity as? BaseActivity)?.showSnackbar(message)
                ?: Log.e(TAG, "Cannot show snackbar: activity does not extend BaseActivity")
    }

    fun getSnackbar(message: String): Snackbar? {
        return (activity as? BaseActivity)?.getSnackbar(message)
    }
}
